File: android/app/src/main/java/com/targetzone/library/data/repository/MembershipRepository.kt
package com.targetzone.library.data.repository

import com.targetzone.library.data.api.ApiClient
import com.targetzone.library.data.model.*

class MembershipRepository {
    private val api = ApiClient.service

    suspend fun getMyMembership(): Result<Membership?> = runCatching {
        val res = api.getMyMembership()
        res.body()?.data
    }

    suspend fun getMembershipHistory(): Result<List<Membership>> = runCatching {
        val res = api.getMembershipHistory()
        res.body()?.data ?: emptyList()
    }

    suspend fun getPlans(): Result<List<Plan>> = runCatching {
        val res = api.getPlans()
        res.body()?.data ?: emptyList()
    }

    suspend fun createOrder(planId: String, seatNumber: String, shift: String): Result<PaymentOrder> = runCatching {
        val res = api.createOrder(CreateOrderRequest(planId = planId, seatNumber = seatNumber, shift = shift))
        res.body()?.data ?: throw Exception(res.body()?.message ?: "Failed to create order")
    }

    suspend fun verifyPayment(req: VerifyPaymentRequest): Result<Membership> = runCatching {
        val res = api.verifyPayment(req)
        res.body()?.data ?: throw Exception(res.body()?.message ?: "Payment verification failed")
    }

    suspend fun submitFeedback(type: String, subject: String, description: String): Result<FeedbackItem> = runCatching {
        val res = api.submitFeedback(SubmitFeedbackRequest(type = type, subject = subject, description = description))
        res.body()?.data ?: throw Exception(res.body()?.message ?: "Failed to submit feedback")
    }

    suspend fun getMyFeedback(): Result<List<FeedbackItem>> = runCatching {
        val res = api.getMyFeedback()
        res.body()?.data ?: emptyList()
    }

    suspend fun getQueuedMembership(): Result<Membership?> = runCatching {
        val res = api.getQueuedMembership()
        res.body()?.data
    }

    suspend fun getMyPayments(): Result<List<StudentPayment>> = runCatching {
        val res = api.getPaymentHistory()
        res.body()?.data ?: emptyList()
    }
}

File: ios/LibrarySystem/LibrarySystem/Features/Student/MembershipRepository.swift
import Foundation

struct MembershipRepository {
    static let shared = MembershipRepository()
    private let api = APIClient.shared
    private var token: String? { TokenManager.shared.token }

    func getMyMembership() async throws -> Membership {
        try await api.request(.getMyMembership, token: token)
    }

    func getMembershipHistory() async throws -> [Membership] {
        try await api.request(.getMembershipHistory, token: token)
    }

    func getQueuedMembership() async -> Membership? {
        try? await api.request(.getQueuedMembership, token: token)
    }

    func getPlans() async throws -> [Plan] {
        try await api.request(.getPlans)
    }

    func createOrder(planId: String, seatNumber: String, shift: String) async throws -> PaymentOrder {
        let req = CreateOrderRequest(planId: planId, seatNumber: seatNumber, shift: shift)
        return try await api.request(.createOrder(req), token: token)
    }

    func verifyPayment(gatewayOrderId: String, gatewayPaymentId: String,
                       signature: String, membershipId: String) async throws -> Membership {
        let req = VerifyPaymentRequest(gatewayOrderId: gatewayOrderId,
                                       gatewayPaymentId: gatewayPaymentId,
                                       signature: signature, membershipId: membershipId)
        return try await api.request(.verifyPayment(req), token: token)
    }

    func submitFeedback(type: String, subject: String, description: String) async throws -> FeedbackItem {
        let req = SubmitFeedbackRequest(type: type, subject: subject, description: description)
        return try await api.request(.submitFeedback(req), token: token)
    }

    func getMyFeedback() async throws -> [FeedbackItem] {
        try await api.request(.getMyFeedback, token: token)
    }

    func getMyPayments() async throws -> [StudentPayment] {
        try await api.request(.getMyPayments, token: token)
    }

    func downloadIdCard() async throws -> Data {
        try await api.download(.downloadIdCard, token: token)
    }
}
